src/rust/demikernel/bindings.rs

```diff
@@ -813,45 +809,101 @@ fn do_syscall<T>(f: impl FnOnce(&mut LibOS) -> T) -> Result<T, Fail> {
     }
 }
 
-/// Converts a [sockaddr] into a [SocketAddrV4].
-fn sockaddr_to_socketaddrv4(saddr: *const sockaddr) -> Result<SocketAddrV4, Fail> {
-    // TODO: Change the logic below and rename this function once we support V6 addresses as well.
-    let sin: SockAddrIn = unsafe {
-        let ptr: *const SockAddrIn = saddr.cast::<SockAddrIn>();
-        ptr.read_unaligned()
-    };
-    if sin.sin_family != AF_INET {
-        return Err(Fail::new(libc::ENOTSUP, "communication domain not supported"));
+/// Converts a [sockaddr] into a [SocketAddr].
+fn sockaddr_to_socketaddr(saddr: *const sockaddr, size: Socklen) -> Result<SocketAddr, Fail> {
+    let check_name_len = |len: usize, exact: bool| {
+        if (size as usize) < len || (exact && size as usize != len) {
+            return Err(Fail::new(libc::EINVAL, "bad socket name length"));
+        }
+        Ok(())
     };
-    let addr: Ipv4Addr = Ipv4Addr::from(u32::from_be(get_addr_from_sock_addr_in(&sin)));
-    let port: u16 = u16::from_be(sin.sin_port);
-    Ok(SocketAddrV4::new(addr, port))
-}
 
-#[test]
-fn test_sockaddr_to_socketaddrv4() {
-    // TODO: assign something meaningful to sa_family and check it once we support V6 addresses as well.
+    // Check that we can read at least the address family from the sockaddr.
+    check_name_len(mem::size_of::<AddressFamily>(), false)?;
 
-    // SocketAddrV4: 127.0.0.1:80
-    #[cfg(target_os = "linux")]
-    let saddr: crate::pal::data_structures::SockAddr = {
-        sockaddr {
-            sa_family: crate::pal::constants::AF_INET_VALUE as u16,
-            sa_data: [0, 80, 127, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
-        }
+    // Read up to size bytes from saddr into a SockAddrStorage, the type which socket2 can use.
+    let mut storage: mem::MaybeUninit<libc::sockaddr_storage> = mem::MaybeUninit::<SockAddrStorage>::zeroed();
```

Comment:
Should this datastructure come from the PAL instead? This would probably not compile on Windows.

Reply:
Good catch; thanks. I let VSCode fill this in, and I guess it "sees through" the typedef.